curvefs/src/volume/block_device_aio.cpp

```diff
@@ -0,0 +1,292 @@
+                 char* data,
+                 FileClient* dev,
+                 int fd)
+    : aio(), offset(offset), length(length), data(data), dev(dev), fd(fd) {}
+
+void AioRead::Issue() {
+    if (is_aligned(offset, IO_ALIGNED_BLOCK_SIZE) &&
+        is_aligned(length, IO_ALIGNED_BLOCK_SIZE)) {
+        aio.ret = -1;
+        aio.op = LIBCURVE_OP_READ;
+        aio.cb = AioReadCallBack;
+        aio.offset = offset;
+        aio.length = length;
+        aio.buf = data;
+
+        int ret = dev->AioRead(fd, &aio);
+        if (ret < 0) {
+            LOG(ERROR) << "Failed to issue aio read: " << &aio;
+            done = true;
+        }
+
+        return;
+    }
+
+    // unaligned read request
+    padding = absl::make_unique<Padding>();
+    auto alignedOff = align_down(offset, IO_ALIGNED_BLOCK_SIZE);
+    auto alignedEnd = align_up(offset + length, IO_ALIGNED_BLOCK_SIZE);
+
+    padding->offset = alignedOff;
+    padding->length = alignedEnd - alignedOff;
+    padding->data.reset(new char[padding->length]);
+
+    aio.ret = -1;
+    aio.op = LIBCURVE_OP_READ;
+    aio.cb = AioReadCallBack;
+    aio.offset = padding->offset;
+    aio.length = padding->length;
+    aio.buf = padding->data.get();
+
+    int ret = dev->AioRead(fd, &aio);
+    if (ret < 0) {
+        LOG(ERROR) << "Failed to issue aio read: " << &aio;
+        done = true;
+    }
+}
+
+ssize_t AioRead::Wait() {
+    std::unique_lock<std::mutex> lock(mtx);
+    cond.wait(lock, [this]() { return done; });
+
+    if (static_cast<ssize_t>(aio.ret) != static_cast<ssize_t>(aio.length)) {
+        LOG(ERROR) << "AioRead error: " << &aio;
+        return -1;
+    }
+
+    if (padding != nullptr) {
+        std::memcpy(data, padding->data.get() + (offset - padding->offset),
+                    length);
+    }
+
+    return length;
+}
+
+AioWrite::AioWrite(off_t offset,
+                   size_t length,
+                   const char* data,
+                   FileClient* dev,
+                   int fd)
+    : offset(offset), length(length), data(data), dev(dev), fd(fd) {}
+
+void AioWrite::Issue() {
+    if (is_aligned(offset, IO_ALIGNED_BLOCK_SIZE) &&
+        is_aligned(length, IO_ALIGNED_BLOCK_SIZE)) {
+        aio.ret = -1;
+        aio.op = LIBCURVE_OP_WRITE;
+        aio.cb = AioWriteCallBack;
+        aio.offset = offset;
+        aio.length = length;
+        aio.buf = const_cast<char*>(data);
+
+        int ret = dev->AioWrite(fd, &aio);
+        if (ret < 0) {
+            LOG(ERROR) << "Failed to issue aio write: " << &aio;
+            done = true;
+        }
+
+        return;
+    }
+
+    aux = absl::make_unique<PaddingAux>();
+
+    aux->alignedOffset = align_down(offset, IO_ALIGNED_BLOCK_SIZE);
+    auto alignedEnd = align_up(offset + length, IO_ALIGNED_BLOCK_SIZE);
+
+    aux->shift = offset - aux->alignedOffset;
+    aux->alignedLength = alignedEnd - aux->alignedOffset;
+
+    aux->paddingData.reset(new char[aux->alignedLength]);
```

Comment:
where free this char[] ?

Reply:
the unique_ptr's destructor will free the memory